src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs

```diff
@@ -0,0 +1,349 @@
+        private static bool TryParseUpdateFile(
+            out string updateFilePath,
+            out SemanticVersion lastUpdateVersion,
+            out DateTime lastUpdateDate)
+        {
+            updateFilePath = null;
+            lastUpdateVersion = null;
+            lastUpdateDate = default;
+
+            var files = Directory.EnumerateFiles(s_cacheDirectory, UpdateFileNamePattern, s_enumOptions);
+            var enumerator = files.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                // It's OK that an update file doesn't exist. This could happen when there is no new updates yet.
+                return true;
+            }
+
+            updateFilePath = enumerator.Current;
+            if (enumerator.MoveNext())
+            {
+                // More than 1 files were found that match the pattern. This is a corrupted state.
+                // Theoretically, there should be only one update file at any point of time.
+                updateFilePath = null;
+                return false;
+            }
+
+            // OK, only found one update file, which is expected.
+            // Now let's parse the file name.
+            string updateFileName = Path.GetFileName(updateFilePath);
+            int dateStartIndex = updateFileName.LastIndexOf('_') + 1;
+
+            if (!DateTime.TryParse(
+                    updateFileName.AsSpan().Slice(dateStartIndex),
+                    CultureInfo.InvariantCulture,
+                    DateTimeStyles.AssumeLocal,
+                    out lastUpdateDate))
+            {
+                updateFilePath = null;
+                return false;
+            }
+
+            int versionStartIndex = updateFileName.IndexOf('_') + 2;
+            int versionLength = dateStartIndex - versionStartIndex - 1;
+            string versionString = updateFileName.Substring(versionStartIndex, versionLength);
+
+            if (SemanticVersion.TryParse(versionString, out lastUpdateVersion))
+            {
+                return true;
+            }
+
+            updateFilePath = null;
+            lastUpdateDate = default;
+            return false;
+        }
+
+        private static async Task<Release> QueryNewReleaseAsync(SemanticVersion baselineVersion)
+        {
+            bool isStableRelease = string.IsNullOrEmpty(PSVersionInfo.PSCurrentVersion.PreReleaseLabel);
+            string queryUri = isStableRelease ? LatestReleaseUri : Last3ReleasesUri;
+
+            using var client = new HttpClient();
+
+            string userAgent = string.Format(CultureInfo.InvariantCulture, "PowerShell {0}", PSVersionInfo.GitCommitId);
+            client.DefaultRequestHeaders.Add("User-Agent", userAgent);
+            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+
+            // Query the GitHub Rest API and throw if the query fails.
+            HttpResponseMessage response = await client.GetAsync(queryUri);
```

Comment:
Should we consider setting `client.Timeout` property?

Reply:
The default timeout for a `HttpClient` instance is 100 seconds, which I think should be fine here.